pkg/storage/chunk/client/baidubce/bos_storage_client.go

```diff
@@ -0,0 +1,173 @@
+package baidubce
+
+import (
+	"context"
+	"flag"
+
+	"io"
+	"time"
+
+	"github.com/baidubce/bce-sdk-go/bce"
+	"github.com/baidubce/bce-sdk-go/services/bos"
+	"github.com/baidubce/bce-sdk-go/services/bos/api"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/instrument"
+
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+)
+
+// ObjectNotFoundErr The specified key does not exist. ref: https://cloud.baidu.com/doc/BOS/s/Ajwvysfpl
+const ObjectNotFoundErr = "ObjectNotFoundErr"
+
+const DefaultEndpoint = "bj.bcebos.com"
+
+var bosRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "loki",
+	Name:      "bos_request_duration_seconds",
+	Help:      "Time spent doing bos requests.",
+	Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
+}, []string{"operation", "status_code"}))
+
+func init() {
+	bosRequestDuration.Register()
+}
+
+type BosStorageConfig struct {
+	BucketName      string `yaml:"bucket_name"`
+	Endpoint        string `yaml:"endpoint"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	SecretAccessKey string `yaml:"secret_access_key"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *BosStorageConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *BosStorageConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.BucketName, prefix+"baidubce.bucket-name", "", "Name of Bos bucket.")
+	f.StringVar(&cfg.Endpoint, prefix+"baidubce.endpoint", DefaultEndpoint, "Bos endpoint to connect to.")
+	f.StringVar(&cfg.AccessKeyID, prefix+"baidubce.access-key-id", "", "Baidu BCE Access Key ID")
+	f.StringVar(&cfg.SecretAccessKey, prefix+"baidubce.secret-access-key", "", "Baidu BCE Secret Access Key")
+}
+
+type BosObjectStorage struct {
+	cfg    *BosStorageConfig
+	client *bos.Client
+}
+
+func (cfg *BosStorageConfig) Validate() error {
+	return nil
```

Comment:
see comment above, do we need to implement this as a no-op to satisfy an interface or can we just avoid calling it?